rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,400 @@
+the client will send a GET request to {dataset_name}/api/v0.01/gene_sets retrieving all of the user's gene sets for
+that dataset.
+
+In the future we will limit the user's ability to create anonymous groups of cells. They will need to use the annotation
+feature to create a category and label the cell group in order to select it for differential expression analysis.
+
+The user will be able to export a CSV file containing one or all of the gene sets they created by selecting cell sets and
+calculating the 10 most differentially expressed genes along with their logfold and p-values between those sets.
+
+The comments section of the csv file will specify the two sets of cells being compared to create that gene set.
+
+The name of the exported file will be based on the user name and a timestamp rounded to the nearest second.
+
+### Keeping Consensus
+
+This feature will require processing multiple updates to the same gene set, due to our hosting architecture this creates
+the potential for race conditions and concerns about keeping the data in sync between the database, server and client.  
+To ensure data consistency, when the user is interacting with the data, the client is always the source of truth.
+If what is stored in the database doesnt match what the user is seeing, the data in the database will be discarded and
+recreated based on the state of the client.
+
+The server will ensure the data is kept in sync through a consensus token or counter. Everytime a gene set is created
+in the database it will have a consensus_counter field set to 0000. This will be sent back to the client along with the
+gene set name and uuid. When the client updates a gene set (via a PUT request) it will send this counter back to the
+server as part of the request body. The server will confirm that the counter sent from the client matches the counter
+stored in the database for that gene set.
+
+- If the values match, the server will update the gene set as specified in the PUT
+  request and update the counter by incrementing it by 1. The response to the client will include the updated counter value
+  to be used in future PUT requests for that gene set.
+- If the counter stored in the database does not match the counter sent by the client the server will immediately send
+  an error message to the client, specifying the name and uuid of the gene set that is out of sync. The client will
+  then send over the entire gene set including all linked genes as a POST request. Because the gene set name is already
+  taken for that user/dataset, the server will delete the currently stored gene set and any associated genes and recreate
+  it based on the information included in the POST request. This will generate a new UUID for the gene set, and reset the
+  consensus counter.
+  In the case that the server times out or does not respond to a PUT request from the client, the client can retry or
+  just send a POST request to recreate the gene set.
```

Comment:
From reading this description I'm not entirely sure what the purpose of the PUT API is, except as a performance optimization.   It looks like the client can use a POST instead.   Since using a PUT has more potential for errors, maybe we should only support the POST?

Reply:
The post is for creating a new gene set, the put is for updates

Reply:
My point is that if the client encounters an error with PUT, then according to the above description, it will then use POST to re-create the gene set.    So, to simplify things, what if the client just started with POST?   In other words when is it beneficial ore preferable to use PUT instead of POST?   
It might be simpler to think of gene sets are immutable.  And they just get replaced if there are ever any changes.  If we do that we can get rid of PUT, and the need for a consensus counter, I think.

Reply:
That would definitely work, and probably be simpler. The thinking here is we will eventually use the same process of incremental updates for annotations as well. This is just an opportunity to try it out and work out any bugs on smaller/easier to debug sets of data.